src/coreclr/nativeaot/System.Private.CoreLib/src/System/Array.NativeAot.cs

```diff
@@ -1196,6 +1199,68 @@ public void RemoveAt(int index)
         }
     }
 
+    internal class SZGenericArrayEnumeratorBase : IDisposable
+    {
+        protected int _index;
+        protected int _endIndex;
+
+        internal SZGenericArrayEnumeratorBase()
+        {
+            _index = -1;
+        }
+
+        public bool MoveNext()
+        {
+            if (_index < _endIndex)
+            {
+                _index++;
+                return (_index < _endIndex);
+            }
+            return false;
+        }
+
+        public void Dispose()
+        {
+        }
+    }
+
+    internal sealed class SZGenericArrayEnumerator<T> : SZGenericArrayEnumeratorBase, IEnumerator<T>
+    {
+        private readonly T[] _array;
+
+        // Passing -1 for endIndex so that MoveNext always returns false without mutating _index
+        internal static readonly SZGenericArrayEnumerator<T> Empty = new SZGenericArrayEnumerator<T>(null, -1);
+
+        internal SZGenericArrayEnumerator(T[] array, int endIndex)
+        {
+            _array = array;
+            _endIndex = endIndex;
+        }
+
+        public T Current
+        {
+            get
+            {
+                if ((uint)_index >= (uint)_endIndex)
+                    ThrowHelper.ThrowInvalidOperationException();
+                return _array[_index];
```

Comment:
This will bounds check on every access, should this maybe use `GetArrayDataReference` and `Unsafe.Add` to avoid doing so?

Reply:
Is this enumerator used often enough on hot paths for it to make a meaningful difference? 

In general, we have stayed away from manual bounds-check elimination using unsafe code in collections.

Reply:
> Is this enumerator used often enough on hot paths for it to make a meaningful difference?

I'd say that arrays are fairly often passed to methods taking `IEnumerable<T>` and this would regress all such usages.

Reply:
> In general, we have stayed away from manual bounds-check elimination using unsafe code in collections.

Looks like we have an explicit bounds check on [line 1244](https://github.com/dotnet/runtime/pull/88371/files#diff-c023b30d1bd19b45a66cfbf46da1f3f9f9cf08b7aaf825d227f8296e4ef8a425R1244) now, so eliminating the one built into array accesses would be safe and worthwhile here.